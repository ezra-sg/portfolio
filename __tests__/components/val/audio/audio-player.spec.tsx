import { act, fireEvent } from '@testing-library/react';

import AudioPlayer, { playbackSpeedOptions } from '@/components/val/audio/audio-player';
import { renderWithLanguage } from '@/__tests__/testing-helpers';

jest.mock('react-icons/md', () => ({
    MdPlayCircleOutline: () => <div>play circle outline</div>,
    MdOutlinePauseCircleOutline: () => <div>pause circle outline</div>,
    MdSpeed: () => <div>speed</div>,
    MdRestartAlt: () => <div>restart alt</div>,
}));

describe('AudioPlayer', () => {
    const pauseAudioSpy = jest.fn();
    const playAudioSpy = jest.fn();

    const AudioPlayerNode = (
        <AudioPlayer
            src="src text"
            labelledBy="labelled by text"
            title="title text"
        />
    );

    beforeAll(() => {
        // address error "Not implemented: HTMLMediaElement.prototype.[pause/play]"
        jest.spyOn(window.HTMLMediaElement.prototype, 'pause')
            .mockImplementation(pauseAudioSpy);
        jest.spyOn(window.HTMLMediaElement.prototype, 'play')
            .mockImplementation(playAudioSpy);
    });

    afterAll(() => {
        jest.restoreAllMocks();
    });

    it('should change states depending on whether the audio is playing, paused, or finished', () => {
        const result = renderWithLanguage(AudioPlayerNode);

        // the 'play' icon should be visible
        expect(result.container).toMatchSnapshot();

        const playButton = result.queryByTestId('audio-player-play-button');
        expect(playButton).toBeInTheDocument();

        act(() => {
            fireEvent.click(playButton!);
        });

        // the 'pause' icon should be visible
        expect(result.container).toMatchSnapshot();

        const audioElement = result.queryByTestId('audio-player-audio-element');
        expect(audioElement).toBeInTheDocument();

        act(() => {
            fireEvent.ended(audioElement!);
        });

        // the 'restart' icon should be visible
        expect(result.container).toMatchSnapshot();
    });

    it('the audio speed menu should work', () => {
        jest.spyOn(document, 'addEventListener');
        jest.spyOn(document, 'removeEventListener');
        const result = renderWithLanguage(AudioPlayerNode);

        // event listener should only be added when the playback speed menu is open
        expect(document.addEventListener).toHaveBeenCalledTimes(0);

        // the playback speed should be 1x by default
        const audioElement = result.queryByTestId('audio-player-audio-element') as HTMLAudioElement;
        expect(audioElement.playbackRate).toBe(1);

        // the playback speed menu should be hidden by default
        const speedMenu = result.queryByTestId('audio-player-speed-menu');
        expect(speedMenu).toBeInTheDocument();
        expect(speedMenu).toHaveAttribute('hidden');

        // open the playback speed menu
        const speedButton = result.queryByTestId('audio-player-speed-button');
        act(() => {
            fireEvent.click(speedButton!);
        });

        expect(speedMenu).not.toHaveAttribute('hidden');

        // clickaway handler should be added to the document when the menu is open
        expect(document.addEventListener).toHaveBeenCalledTimes(1);
        const clickawayHandler = (document.addEventListener as jest.Mock).mock.calls[0][1];

        const speedOptionOne = result.queryByTestId('audio-player-speed-option-0');
        expect(speedOptionOne).toBeInTheDocument();
        act(() => {
            // click the first playback speed option
            fireEvent.click(speedOptionOne!);
        });

        // the playback speed should be updated
        expect(audioElement.playbackRate).toBe(playbackSpeedOptions[0]);
        expect(speedMenu).toHaveAttribute('hidden'); // selecting an option should close the menu
        expect(document.removeEventListener).toHaveBeenCalledTimes(1); // clickaway handler should be removed when the menu is closed

        // re-open the playback speed menu
        act(() => {
            fireEvent.click(speedButton!);
        });

        act(() => {
            // fireEvent/userEvent.click on the document doesn't trigger event handlers
            // so we need to call the handler directly to simulate a click outside the menu
            clickawayHandler({ target: document.body });
        });

        // the playback speed menu should be hidden on clickaway
        expect(speedMenu).toHaveAttribute('hidden');
        expect(document.removeEventListener).toHaveBeenCalledTimes(2);
        expect(document.removeEventListener).toHaveBeenCalledWith('mousedown', clickawayHandler);
    });
});
