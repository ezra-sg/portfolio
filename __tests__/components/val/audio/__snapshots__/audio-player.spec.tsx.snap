// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer should change states depending on whether the audio is playing, paused, or finished 1`] = `
<div>
  <div
    class="flex items-center gap-1 text-2xl text-amber-900 dark:text-orange-300"
  >
    <button
      aria-label="inputs.play_audio_aria title text"
      class="h-8 w-8 flex items-center justify-center"
      data-testid="audio-player-play-button"
      title="inputs.play_audio_aria title text"
    >
      <div>
        play circle outline
      </div>
    </button>
    <code
      class="text-xs"
      title="inputs.audio_time_elapsed_title title text"
    >
      00:00
    </code>
    <input
      aria-label="inputs.audio_scrubber_aria title text"
      class="c-audio-player__scrubber"
      min="0"
      step="0.1"
      type="range"
      value="0"
    />
    <code
      class="text-xs"
      title="inputs.audio_time_total_title title text"
    >
      00:00
    </code>
    <div
      class="relative"
    >
      <button
        aria-label="inputs.change_playback_speed_aria title text"
        class="h-8 w-8 flex items-center justify-center"
        data-testid="audio-player-speed-button"
        title="inputs.change_playback_speed_aria title text"
      >
        <div>
          speed
        </div>
      </button>
      <ul
        aria-hidden="true"
        aria-label="inputs.audio_speed_menu_label title text"
        class="absolute p-3 bg-amber-50 shadow-lg rounded-sm border-[1px] border-amber-900 dark:bg-stone-950 dark:border-orange-300"
        data-testid="audio-player-speed-menu"
        hidden=""
        role="menu"
      >
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.25"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-0"
          role="menuitemradio"
          tabindex="0"
        >
          0.25x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.5"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-1"
          role="menuitemradio"
          tabindex="0"
        >
          0.5x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.75"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-2"
          role="menuitemradio"
          tabindex="0"
        >
          0.75x
        </li>
        <li
          aria-checked="true"
          aria-label="inputs.audio_speed_item_label 1"
          class="font-bold text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-3"
          role="menuitemradio"
          tabindex="0"
        >
          global.normal
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.25"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-4"
          role="menuitemradio"
          tabindex="0"
        >
          1.25x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.5"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-5"
          role="menuitemradio"
          tabindex="0"
        >
          1.5x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.75"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-6"
          role="menuitemradio"
          tabindex="0"
        >
          1.75x
        </li>
      </ul>
    </div>
  </div>
  <audio
    aria-labelledby="labelled by text"
    controls=""
    data-testid="audio-player-audio-element"
    hidden=""
    src="src text"
  />
</div>
`;

exports[`AudioPlayer should change states depending on whether the audio is playing, paused, or finished 2`] = `
<div>
  <div
    class="flex items-center gap-1 text-2xl text-amber-900 dark:text-orange-300"
  >
    <button
      aria-label="inputs.pause_audio_aria title text"
      class="h-8 w-8 flex items-center justify-center"
      data-testid="audio-player-play-button"
      title="inputs.pause_audio_aria title text"
    >
      <div>
        pause circle outline
      </div>
    </button>
    <code
      class="text-xs"
      title="inputs.audio_time_elapsed_title title text"
    >
      00:00
    </code>
    <input
      aria-label="inputs.audio_scrubber_aria title text"
      class="c-audio-player__scrubber"
      min="0"
      step="0.1"
      type="range"
      value="0"
    />
    <code
      class="text-xs"
      title="inputs.audio_time_total_title title text"
    >
      00:00
    </code>
    <div
      class="relative"
    >
      <button
        aria-label="inputs.change_playback_speed_aria title text"
        class="h-8 w-8 flex items-center justify-center"
        data-testid="audio-player-speed-button"
        title="inputs.change_playback_speed_aria title text"
      >
        <div>
          speed
        </div>
      </button>
      <ul
        aria-hidden="true"
        aria-label="inputs.audio_speed_menu_label title text"
        class="absolute p-3 bg-amber-50 shadow-lg rounded-sm border-[1px] border-amber-900 dark:bg-stone-950 dark:border-orange-300"
        data-testid="audio-player-speed-menu"
        hidden=""
        role="menu"
      >
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.25"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-0"
          role="menuitemradio"
          tabindex="0"
        >
          0.25x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.5"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-1"
          role="menuitemradio"
          tabindex="0"
        >
          0.5x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.75"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-2"
          role="menuitemradio"
          tabindex="0"
        >
          0.75x
        </li>
        <li
          aria-checked="true"
          aria-label="inputs.audio_speed_item_label 1"
          class="font-bold text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-3"
          role="menuitemradio"
          tabindex="0"
        >
          global.normal
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.25"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-4"
          role="menuitemradio"
          tabindex="0"
        >
          1.25x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.5"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-5"
          role="menuitemradio"
          tabindex="0"
        >
          1.5x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.75"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-6"
          role="menuitemradio"
          tabindex="0"
        >
          1.75x
        </li>
      </ul>
    </div>
  </div>
  <audio
    aria-labelledby="labelled by text"
    controls=""
    data-testid="audio-player-audio-element"
    hidden=""
    src="src text"
  />
</div>
`;

exports[`AudioPlayer should change states depending on whether the audio is playing, paused, or finished 3`] = `
<div>
  <div
    class="flex items-center gap-1 text-2xl text-amber-900 dark:text-orange-300"
  >
    <button
      aria-label="inputs.restart_audio_aria title text"
      class="h-8 w-8 flex items-center justify-center"
      data-testid="audio-player-play-button"
      title="inputs.restart_audio_aria title text"
    >
      <div>
        restart alt
      </div>
    </button>
    <code
      class="text-xs"
      title="inputs.audio_time_elapsed_title title text"
    >
      00:00
    </code>
    <input
      aria-label="inputs.audio_scrubber_aria title text"
      class="c-audio-player__scrubber"
      min="0"
      step="0.1"
      type="range"
      value="0"
    />
    <code
      class="text-xs"
      title="inputs.audio_time_total_title title text"
    >
      00:00
    </code>
    <div
      class="relative"
    >
      <button
        aria-label="inputs.change_playback_speed_aria title text"
        class="h-8 w-8 flex items-center justify-center"
        data-testid="audio-player-speed-button"
        title="inputs.change_playback_speed_aria title text"
      >
        <div>
          speed
        </div>
      </button>
      <ul
        aria-hidden="true"
        aria-label="inputs.audio_speed_menu_label title text"
        class="absolute p-3 bg-amber-50 shadow-lg rounded-sm border-[1px] border-amber-900 dark:bg-stone-950 dark:border-orange-300"
        data-testid="audio-player-speed-menu"
        hidden=""
        role="menu"
      >
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.25"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-0"
          role="menuitemradio"
          tabindex="0"
        >
          0.25x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.5"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-1"
          role="menuitemradio"
          tabindex="0"
        >
          0.5x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 0.75"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-2"
          role="menuitemradio"
          tabindex="0"
        >
          0.75x
        </li>
        <li
          aria-checked="true"
          aria-label="inputs.audio_speed_item_label 1"
          class="font-bold text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-3"
          role="menuitemradio"
          tabindex="0"
        >
          global.normal
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.25"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-4"
          role="menuitemradio"
          tabindex="0"
        >
          1.25x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.5"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-5"
          role="menuitemradio"
          tabindex="0"
        >
          1.5x
        </li>
        <li
          aria-checked="false"
          aria-label="inputs.audio_speed_item_label 1.75"
          class=" text-sm cursor-pointer my-2 hover:underline"
          data-testid="audio-player-speed-option-6"
          role="menuitemradio"
          tabindex="0"
        >
          1.75x
        </li>
      </ul>
    </div>
  </div>
  <audio
    aria-labelledby="labelled by text"
    controls=""
    data-testid="audio-player-audio-element"
    hidden=""
    src="src text"
  />
</div>
`;
